### ISSUE TITLE
New pause command

### ISSUE THREAD

**Issue by bob2517** [Active-CSS/active-css#155]
With the synchronous "await" evolution for 2.6.0, in practice, chaining regular action commands using await can get a little bit confusing.

So I'm creating a pause command, which separates the delay into an actual command.

Essentially, the test code looks easier to read with a pause command, like:

```
div:draw {
    background-color: blue;
    pause: 1s;
    background-color: green;
    pause: 1s;
    background-color: blue;
}
```

when compared to:

```
div:draw {
    background-color: blue;
    background-color: green after 1s await;
    background-color: blue after 1s;
}
```

which does the same thing, but requires more brainpower to work out what's going on, even if it is quicker to type.

This can also be written like:

```
div:draw {
    background-color: blue, green after 1s await, blue after 1s;
}
```

So the await command has power, but can arguably be a bit harder to read.

The await option therefore could make more sense for ajax-type commands. But it is handy for one-off delays, so I'm keeping it in the core.

The pause command will internally be asynchronous, like await, but from a developer viewpoint will be synchronous within a specific event declaration. Pausing shouldn't require the whole universe to stop, like in JavaScript, when it is used, even if JS uses the excuse of being single-threaded. The monitoring of time is done on a single entity or particle anyway, and so doing it this way more mirrors the way time is used in the physical universe, and so more accurately matches developer intent when they want to pause. Ie., when you want to stop moving your hand, the rest of your body doesn't have to stop. Having synchronous pauses within a specific event/element time stream therefore makes sense.

A step further on from this would be a way to asynchronously chain the same action command, so the same action command can run multiple times being chained, but allow other multiple chained commands to run at same time in the same event. Eg. color can also change at the same time something changes position, within the same event. It can be done with the existing syntax using "after", but a proper syntax for it would be cool. Maybe like a "chain" option which chains the same command if it is followed by the same command, instead of "await". Would need to work that out though. Like:

```
div:draw {
    background-color: blue, green after 1s chain, blue after 1s;
    color: green, blue after 1s chain, green after 1s;
```

In this case both commands would start at the same time, but each command is chained, something like that. Still not easy to read like that though...

I'll get this evolution wrapped up first and then ponder that last one for a future release.

**Comment by bob2517** [comment]
This is almost done, BTW, but I found an error when used with variables and loops, as internally the code isn't in the right place. So this weekend I'm going to jig things around a little bit to sort that out.

It would be good to have pause-transition/resume-transition and pause-animation/resume-animation commands at some point, but that's a completely different thing and nothing to do with the action of the pause command or the await option...

**Comment by bob2517** [comment]
Doing this has given me an idea for restarting an event sequence rather than re-triggering the outer event. Gonna pend this until I think of a use case. But if anyone needs it it should be easy enough to implement after doing this issue.

**Comment by bob2517** [comment]
Gah. Need to change implementation method completely. There's only one sensible way to get this working with loops, and that's never to leave the loop at all, which isn't the way I've set it up.

Re-writing this now. Just need to work out how the ajax commands will work with this...

**Comment by bob2517** [comment]
The only way I can do this sensibly is by using delay promises after direct action command calls, and that will only work if each event has its own event "thread" by using setTimeout. This will break stop-event-propagation and stop-immediate-event-propagation commands, so I'm going to set it up so that these new threads will only be created if await or pause commands are in the event. That way those commands can still be useful and existing configs won't suddenly do something different. It doesn't change the "after" option on action commands, as that gave a specific action command its own thread in much the same way, but at a lower level.

**Comment by bob2517** [comment]
Just need to work out how promisify XHR and that should be it for the theory. Can't use fetch as I need the flexibility of XHR for future stuff planned.

**Comment by bob2517** [comment]
Meh. Doesn't look like that's going to work for whatever reason. I'm probably getting too complicated - gonna simplify my thinking a little bit and try something more basic...

**Comment by bob2517** [comment]
Double meh. Async/await isn't ES6...

**Comment by bob2517** [comment]
Experimenting with using a promise in loop. Looks like it could work maybe.

**Comment by bob2517** [comment]
Yeah, that's not going to work either. Looks like I'm going to have to write a manual iterator. ES8 sounds promising though in about 10 years when I can actually use it without Babel.

**Comment by bob2517** [comment]
I might use a function generator - not had a reason to use that yet. I like the idea of the next() thing.

**Comment by bob2517** [comment]
The good thing about an iterator that I have to write myself though is that it will work easily with ajax-type commands and I can resume iterations anywhere in the core that I like, so I suppose that is the plus side.

**Comment by bob2517** [comment]
I need to do some serious surgery on the core for this to work. All the for loops below the event level need to stop being for loops and become manual iterators of some kind, with manual "callbacks". That's the only way I can get asynchronicity to the level I need for everything to work correctly with loops and variables. If I didn't have to cater for loops and variables, and even ES6 (i could have used async/await) then this problem wouldn't have come up. I'm not going to add Babel at this point, as it always adds on bloat.

It's either that or get a sensible polyfill for async/await, but I'm doubtful now that a good enough solution is available for what I need.

So tomorrow, I'm going to convert all the for loops starting from the top level function above the loops, so that each can be manually iterated with callbacks, just on the core as it stands live at the moment. I'll get that working. Then add in the settimeouts pointing to those callbacks to prompt the next iteration of whatever, and that should do the job. Then the next loop iterations can be run from anywhere. I might set up a generic iterator function of some kind to simplify things a bit - maybe not though.

**Comment by bob2517** [comment]
Switching strategy again. For simplicity's sake I'm gonna use babel so I can use async/await. It's not worth rewriting everything for this. Core would get larger if I rewrote anyway. Peeps who need to support anything earlier than 2017 can use the babel version. I'm going to stick with ES6 for the rest, unless I get into a position like this again.

I think async/await should do the job. Will need to convert XHR for it and change the existing callback strategies, but will give this a go. Rewriting is a last resort. I'd rather switch to fetch and have to get a different workaround for future implementations that would have needed XHR than do a rewrite just because of XHR.

**Comment by bob2517** [comment]
Lol, async/await isn't going to cut the cheese either. I thought you'd get true synchronous and be able to mix async functions with sync functions but it's not possible to keep the event flow intact using it, which was the whole point. Going back to my original plan of rewriting the core. I'll put the core back to it all working with await and pause as they were prior to the loop fixes and rewrite it piece by from the top down, getting it all to work, probably by removing the loops and doing stuff with shift(). Then I'll be able to move the await and pause code to their correct places easily. At least I can get rid of babel too - it added 14kb onto the min core just by using it to support async/await.

**Comment by bob2517** [comment]
The way I'm doing this: 1) convert loops in all key event flow functions into recursive functions so the loops are totally gone. 2) From recursive functions, break up all the recursions by converting to manual function calls. 3) With manual function calls acting as callbacks, rejig the await/pause code. In theory, with manual callbacks I should retain control of the stack and have as much asynchronous stuff going on as I want. I can't see a better way of doing it and maintaining the event flow and keeping the performance benefits by using as little async code as possible. It's basically just one controlled flow. I just need to make sure I'm passing enough data around to keep everything working. I'm about half way through step 1.

**Comment by bob2517** [comment]
Step 1 is done, and everything works so far with the loops gone. Now onto step 2.

**Comment by bob2517** [comment]
Weirdly, GOTO statements would be quite handy at this point. It's a prehistoric technique I'm having to employ to get around the problems with async destroying the event flow. Let's hope it works, which I'll find out soon enough. If it doesn't then at least the loops are gone, which isn't a bad thing.

**Comment by bob2517** [comment]
It's looking good so far. I've set up a manual iteration at the bottom of the event flow, and I think I can put all my manual redirection code in one function, as the whole point of flow is to run an action command. So I think I can just run one function at the point of action command handling to cater for all positions in the flow and remove all the higher iteration steps that would have been ran into after returns. Dunno if that will pan out, but it's looking like it could happen which would be cool. Just need to make sure the higher objects carry down the chain untampered with so that properties can determine where to go. Will continue tomorrow.

**Comment by bob2517** [comment]
Use some sort of resumption object containing start positions in iterations and the object each applies to. Then call the highest level above the loop with the original object result set. Skip ones below the steps in the resumption object and when there's a match drill down and repeat. That's probably the most sensible way the event loop structure can be restored to what it was before the await and keep in its dynamic nature. Using an event queue like before won't work as it isn't dynamic.

**Comment by bob2517** [comment]
In fact, the changes made yesterday to remove the loops could be removed, as it isn't necessary to do this if using a resumption object as the loops must start from the beginning anyway and it doesn't matter. Maybe set up a different branch and have a play around. Remove changes if the core is now harder to read since Friday.

**Comment by bob2517** [comment]
When handling the variable looping resumption, bear in mind that there may have been changes to variables during the await or pause via a different event. The resumption needs to look out for these changes and either skip or break out if a right hand variable no longer matches, or more drastically restore to the original value before the change, or something. Work out which to do when you get to it by using a test case.

**Comment by bob2517** [comment]
Actually - leave the code split up as it is. It's probably easier for other people to get an idea of how it works, as it now has more defined sections.

**Comment by bob2517** [comment]
So next steps in sequence (I can't be on this all day unfortunately as I've got other stuff going on):

1) Create the resumption object.
2) Check the values of the resumption object at the point of await detection, which will be in _handleFunc at the bottom of the chain.
3) Get the code not performing any more actions once await has been started (it does this already - just check it).
4) Make sure the variables necessary for resumption in restarting the higher-level loop function are comparable to the initial  values and available after the settimeout.
5) Call the higher-level function.
At this point, the whole event loop should be restarted from the beginning. It should work but not how we need.
6) Insert conditional checks at the top of each loop and check if the counter is below what we need - if so, increment and re-call the function. Use a generic comparison function for this.
7) When it comes to checking the objects... just don't start drilling into the loop until we find the object. We can have a rule that if the paused object gets removed during a loop, the await process stops. Implement that for now. Maybe come back and check later if a different solution is discovered to be wanted. Maybe an option on await could determine what to do if the loop is allowed to continue after an object is remove. Anyway - for now keep looking for the originally paused object in order to continue the loop. This at least gives dynamic action to elements that the pause didn't occur on. It won't work on new elements that got added subsequent to the paused loop, but that's wanted behaviour.
8) Then handle the loop variables as in the above comment.

That should be all the steps and allow regular settimeouts to pause and resume where the event queue left off at the level high enough to respond dynamically.

**Comment by bob2517** [comment]
If mid a target selector set while paused or asynced, the original result set will be used to finish off. With that in mind, the earlier working version which kept an event queue could be used to keep the original result sets.
Dynamic element selector or using the elements available when the event started - that is the question. I think the more predictable option is to use the elements that would have been available when the event started. But the future implementation of having loops around target selectors dictates that I won't know what the target selectors are until the loops are began. And I can't begin the loops as variables will get set at the wrong time. I think go with the dynamic option and see how it goes. The dynamic option also makes sense logically. It depends what you want to do on the page I guess.

**Comment by bob2517** [comment]
Further notes on this to avoid weird bugs coming up later. Points 1 to 5 are done above, so into the meat of the resumption next.

When resuming the event flow, the elements involved in target selector need to remain exactly the same as prior to the pause. Do this by storing each selector set (to be safe, this should be per loop sequence) and using these instead of new element queries. The same for variables. Variables shouldn't change mid a loop as it isn't good form, but in case they do the original right-hand variable needs to be used instead of the dynamic one. Otherwise someone might resume after a pause and get very unexpected results.

Essentially, the resumption needs to be an exact duplicate of the original event. It then doesn't matter if target selectors are no longer there, as the elements will be referenced but just no longer connected to the DOM. Once the correct point has been reached in the event flow to resume, everything can be put back to normal so it works dynamically from that point as would be expected.

**Comment by bob2517** [comment]
Resumption points to detect during the loop: target selector, left-hand variables in loop(s), unique action command ID from the config (if there is one - if not, add it). All those are required to spot the correct point paused.

**Comment by bob2517** [comment]
This is pretty much done. Just need to write some more tests, etc. Might do a commit later on today. The above method works and seem to give a true synchronous effect (which is no mean feat considering that JavaScript doesn't handle separate event stacks natively). But more complex tests are definitely needed just to be on the safe side before release.

**Comment by bob2517** [comment]
Sneak peak:
![image](https://user-images.githubusercontent.com/42515015/120495413-9833e700-c3b4-11eb-95be-4e86c5b6510a.png)

**Comment by bob2517** [comment]
Example syntax:

```
span:draw {
	pause: 1s;	
	background-color: #{$RANDHEX6};
	pause: 1s;	
	border-radius: 50%;
	pause: 1s;	
	background-color: #{$RANDHEX6};
	pause: 1s;	
	border-radius: 0%;
	trigger: draw;
}
```

**Comment by bob2517** [comment]
This also works in much the same way, which pauses after each delayed background change:

```
span:draw {
    background-color: #{$RANDHEX6} after 1s await, #{$RANDHEX6} after 1s await, #{$RANDHEX6} after 1s await, #{$RANDHEX6} after 1s await, #{$RANDHEX6} after 1s await;
    trigger: draw;
}
```

As does this:

```
button:click {
    ajax: "/get-address-data.php" post-pars(id=2) JSON await;
    render-after-end: "<div>{title}<br>{address}</div>";
    ajax: "/get-address-data.php" post-pars(id=3) JSON await;
    render-after-end: "<div>{title}<br>{address}</div>";
}

button:afterAjaxError {
	render-replace: "<p>Error loading address.</p>";
}
```

**Comment by bob2517** [comment]
I'm so happy with this - it's such a relief to finally crack synchronous coding in an asynchronous way. I'm almost tempted to skip the tests. Just kidding. Or am I?

**Comment by bob2517** [comment]
This now works using the new method, which is good as it wasn't working with the old method:

```
span:draw {
	var: test ['bob', 'dave', 'trevor'];
	@each item in test {
		render-before-end: "{item}<br>";
		pause: 1s;
	}
}
```

**Comment by bob2517** [comment]
This works too. Oh yeah baby...

```
span:draw {
	var: test2 ['hi', 'goodbye'], test ['bob', 'dave', 'trevor'];
	@each item2 in test2 {
		@each item in test {
			render-before-end: "{item2} {item}<br>";
			pause: 1s;
		}
	}
}
```

**Comment by bob2517** [comment]
Just need to update the docs and convert these examples into proper tests.

Marking as done offline... :) :) :) :) :) :) :) :) :) :) :) :) :) :) :) :) :) :)

**Comment by bob2517** [comment]
An awesome thing about this method in the core is that it is really easy to set up for any new action command. It just needs a regular settimeout and couple of function calls and that's it.

**Comment by bob2517** [comment]
And it will automatically work with the future @for loop syntax and also allowing loops to be around action commands. Which is good too. All the resumption-after-pausing code is right at the bottom at the function level, and everything above it can get as weird as it likes.

**Comment by bob2517** [comment]
Having done this, it really highlights the need for being able to wrap loops around action commands. Like you can't currently do this because the first \@each is around an action command:

```
span:draw {
    var: test2 ['hi', 'goodbye'], test ['bob', 'dave', 'trevor'];
    @each item2 in test2 {
        pause: 2s;
        @each item in test {
            render-before-end: "{item2} {item}<br>";
            pause: 250ms;
        }
    }
}
```

2.6.0 await/pause is quite an achievement in itself though, so I'm going to release 2.6.0 as it is and do the loop wrap-up for 2.7.0.

**Comment by bob2517** [comment]
Minor memory leak on this, which doesn't surprise me, so will have that fixed at the weekend when I'm back on this.

**Comment by bob2517** [comment]
Back on this. Pushing for release today!

**Comment by bob2517** [comment]
Just been spending the last few hours (around the grand prix) trying to fix a bug that was more like something that happens as a result of a pause.

Consider this on a set of 4 span tags:

```
button:click {
    span:draw {
         background-color: blue;
         pause: 1s;
         background-color: green;
         pause: 1s;
    }
    div {
        render: "Finished.";
    }
}
```

What should happen? Well, you might think that all the spans would work simultaneously, but not so. Because action commands are sequential per event, the first span runs to the first pause, then the second span runs to the first pause, then the third span, etc. Then the first span runs up to the second pause, the second span runs up to the second pause, etc. When all spans have finished their last pauses, the div renders "Finished."

To get around this, the following can be done which triggers a custom event on each of the spans:

```
button:click {
    span {
        trigger: customEvent;
    }
}

span:customEvent {
     background-color: blue;
     pause: 1s;
     background-color: green;
     pause: 1s;
}
```

I'm not convinced this is right yet. Ideally I would expect all multiple set target selectors to set off simultaneous events with simultaneous pauses, and then when all are completed, move onto the next target selector.

I could just push this live with the quirk, but I think I'm going to try and get this sorted out.

I think what it does at the moment is a valid use case and does actually make logical sense. It's the way the core has been running all along - one target selector in a set would run fully, then the next, etc. But it's not what I expected in testing this scenario, so it depends on the viewpoint of the developer and what he needs.

So I'm going to get an option of some kind to set the way target selectors behave with pause and try and work out a way to get some sort of simultaneous target selector thing going on. I think then all the scenarios should be covered.

**Comment by bob2517** [comment]
Been wrestling with tests today. I'm happy with this now. "Await" now has two actions. if "Await" is applied to multiple target selectors, it cycles through each target selector applying the pause as it goes, so you can get a sequenced linear effect across multiple elements. But if applied to something like a draw event, which can run multiple events, then you can get a simultaneous effect. You can also trigger off an event if you need to apply a simultaneous effect to target selectors. I've optimised the flow today, so performance is as lean as it can be with the current setup.

I think I'm going to go with the end result I've got now - it's pretty good - you can do two things and mix things up depending on the technique that is used, so it looks quite neat. As long as you understand how the event flow works, it all makes sense. So I need to update the documentation and see how far I get this evening. I may just push it live tomorrow afternoon though.

**Comment by bob2517** [comment]
Closing - live on 2.6.0.
